File: feishu_service.js
import { FEISHU_CONFIG } from './config.js';

export class FeishuService {
    constructor() {
        this.baseUrl = "https://open.feishu.cn";
        this.appAccessToken = null;
    }

    setCredentials(appId, appSecret) {
        this.appId = appId;
        this.appSecret = appSecret;
    }

    async makeRequest(url, options) {
        throw new Error('makeRequest not implemented');
    }

    async getAppAccessToken() {
        console.group('获取 app_access_token');
        try {
            const result = await this.makeRequest(
                `${this.baseUrl}/open-apis/auth/v3/app_access_token/internal`,
                {
                    method: 'POST',
                    body: {
                        app_id: this.appId,
                        app_secret: this.appSecret
                    }
                }
            );

            if (result.code === 0) {
                this.appAccessToken = result.app_access_token;
                return this.appAccessToken;
            }
            throw new Error(`获取app_access_token失败: ${JSON.stringify(result)}`);
        } catch (error) {
            console.error('错误:', error);
            throw error;
        } finally {
            console.groupEnd();
        }
    }

    async getUserAccessToken(appAccessToken, code) {
        try {
            const result = await this.makeRequest(
                `${this.baseUrl}/open-apis/authen/v1/access_token`,
                {
                    method: 'POST',
                    headers: {
                        'Authorization': `Bearer ${appAccessToken}`,
                        'Content-Type': 'application/json',
                        'Accept': 'application/json'
                    },
                    body: JSON.stringify({
                        grant_type: "authorization_code",
                        code: code
                    })
                }
            );
            
            if (result.code === 0) {
                console.log("userAccessToken", result.data);
                const refreshedTokenData = await this.refreshUserAccessToken(result.data.access_token, result.data.refresh_token);
                return refreshedTokenData;
            } else {
                throw new Error(`获取user_access_token失败: ${JSON.stringify(result)}`);
            }
        } catch (error) {
            throw new Error(`获取user_access_token请求失败: ${error.message}`);
        }
    }

    async refreshUserAccessToken(userAccessToken, refreshToken) {
        try {
            const result = await this.makeRequest(
                `${this.baseUrl}/open-apis/authen/v1/refresh_access_token`,
                {
                    method: 'POST',
                    headers: {
                        'Authorization': `Bearer ${userAccessToken}`,
                        'Content-Type': 'application/json',
                        'Accept': 'application/json'
                    },
                    body: JSON.stringify({
                        grant_type: "refresh_token",
                        refresh_token: refreshToken
                    })
                }
            );

            if (result.code === 0) {
                console.log("refreshedToken", result.data);
                return result.data;
            } else {
                throw new Error(`刷新user_access_token失败: ${JSON.stringify(result)}`);
            }
        } catch (error) {
            throw new Error(`刷新user_access_token请求失败: ${error.message}`);
        }
    }

    async uploadFile(fileData, fileName, userAccessToken, parentNode = "") {
        try {
            const formData = new FormData();
            const blob = new Blob([fileData], { type: 'text/markdown' });
            
            formData.append('file_name', fileName);
            formData.append('parent_type', 'explorer');
            if (parentNode) {
                formData.append('parent_node', parentNode);
            }
            formData.append('size', blob.size.toString());
            formData.append('file', blob, fileName);

            const result = await this.makeRequest(
                `${this.baseUrl}/open-apis/drive/v1/files/upload_all`,
                {
                    method: 'POST',
                    headers: {
                        'Authorization': `Bearer ${userAccessToken}`,
                        'Accept': 'application/json'
                    },
                    body: formData
                }
            );

            if (result.code === 0) {
                return result.data;
            } else {
                throw new Error(`上传文件失败: ${JSON.stringify(result)}`);
            }
        } catch (error) {
            throw new Error(`上传文件请求失败: ${error.message}`);
        }
    }

    getOAuthUrl() {
        const url = new URL(`${this.baseUrl}/open-apis/authen/v1/index`);
        url.searchParams.append('client_id', this.appId);
        url.searchParams.append('redirect_uri', FEISHU_CONFIG.REDIRECT_URI);
        url.searchParams.append('scope', FEISHU_CONFIG.SCOPE);
        url.searchParams.append('state', FEISHU_CONFIG.STATE);
        console.log("url", url.toString());
        return url.toString();
    }
} 

File: feishu_service.py
import requests
import json
from typing import Optional

class FeishuService:
    def __init__(self):
        self.base_url = "https://open.feishu.cn"
        self.app_access_token = None
        
    def set_credentials(self, app_id: str, app_secret: str):
        """设置应用凭证"""
        self.app_id = app_id
        self.app_secret = app_secret
    
    def get_app_access_token(self) -> str:
        """获取 app_access_token"""
        url = f"{self.base_url}/open-apis/auth/v3/app_access_token/internal"
        payload = {
            "app_id": self.app_id,
            "app_secret": self.app_secret
        }
        
        response = requests.post(url, json=payload)
        result = response.json()
        
        if result.get("code") == 0:
            self.app_access_token = result.get("app_access_token")
            return self.app_access_token
        else:
            raise Exception(f"获取app_access_token失败: {result}")

    def get_oauth_url(self, redirect_uri: str) -> str:
        """生成飞书授权链接"""
        return (f"{self.base_url}/open-apis/authen/v1/index"
                f"?app_id={self.app_id}"
                f"&redirect_uri={redirect_uri}")
    
    def get_user_access_token(self, code: str) -> dict:
        """通过授权码获取user_access_token"""
        url = f"{self.base_url}/open-apis/authen/v1/access_token"
        payload = {
            "grant_type": "authorization_code",
            "code": code
        }
        headers = {
            "Authorization": f"Bearer {self.app_access_token or self.get_app_access_token()}"
        }
        
        response = requests.post(url, json=payload, headers=headers)
        result = response.json()
        
        if result.get("code") == 0:
            return result.get("data")
        else:
            raise Exception(f"获取user_access_token失败: {result}")
    
    def refresh_user_access_token(self, refresh_token: str) -> dict:
        """刷新user_access_token"""
        url = f"{self.base_url}/open-apis/authen/v1/refresh_access_token"
        payload = {
            "grant_type": "refresh_token",
            "refresh_token": refresh_token
        }
        headers = {
            "Authorization": f"Bearer {self.app_access_token or self.get_app_access_token()}"
        }
        
        response = requests.post(url, json=payload, headers=headers)
        result = response.json()
        
        if result.get("code") == 0:
            return result.get("data")
        else:
            raise Exception(f"刷新user_access_token失败: {result}")
    
    def upload_file(self, file_data: bytes, file_name: str, user_access_token: str) -> dict:
        """上传文件到飞书云空间"""
        url = f"{self.base_url}/open-apis/drive/v1/files/upload_all"
        
        files = {
            'file': (file_name, file_data)
        }
        headers = {
            "Authorization": f"Bearer {user_access_token}"
        }
        
        response = requests.post(url, files=files, headers=headers)
        result = response.json()
        
        if result.get("code") == 0:
            return result.get("data")
        else:
            raise Exception(f"上传文件失败: {result}") 